src/core/or/circuitpadding.h

```diff
@@ -0,0 +1,539 @@
+   * This is the recommended strategy from the Adaptive Padding paper. */
+  CIRCPAD_TOKEN_REMOVAL_HIGHER = 1,
+  /** Remove from the first non-zero lower bin index when current is empty. */
+  CIRCPAD_TOKEN_REMOVAL_LOWER = 2,
+  /** Remove from the closest non-zero bin index when current is empty. */
+  CIRCPAD_TOKEN_REMOVAL_CLOSEST = 3,
+  /** Remove from the closest bin by time value (since bins are
+   *  exponentially spaced). */
+  CIRCPAD_TOKEN_REMOVAL_CLOSEST_USEC = 4,
+  /** Only remove from the exact bin corresponding to this delay. If
+   *  the bin is 0, simply do nothing. Don't pick another bin. */
+  CIRCPAD_TOKEN_REMOVAL_EXACT = 5
+} circpad_removal_t;
+
+/** The maximum length any histogram can be. */
+/** Each bin is twice as large as the previous, with an extra bin for 0 */
+#define CIRCPAD_MAX_HISTOGRAM_LEN (sizeof(circpad_delay_t)*8 + 1)
+
+/**
+ * Distribution types supported by circpad_distribution_sample().
+ *
+ * These can be used instead of histograms for the inter-packet
+ * timing distribution, or to specify a distribution on the number
+ * of cells that can be sent while in a specific state of the state
+ * machine. */
+typedef enum {
+  CIRCPAD_DIST_NONE = 0,
+  CIRCPAD_DIST_UNIFORM = 1,
+  CIRCPAD_DIST_LOGISTIC = 2,
+  CIRCPAD_DIST_LOG_LOGISTIC = 3,
+  CIRCPAD_DIST_GEOMETRIC = 4,
+  CIRCPAD_DIST_WEIBULL = 5,
+  CIRCPAD_DIST_PARETO = 6
+} circpad_distribution_type_t;
+
+/**
+ * Distribution information.
+ *
+ * This type specifies a specific distribution above, as well as
+ * up to two parameters for that distribution. The specific
+ * per-distribution meaning of these parameters is specified
+ * in circpad_distribution_sample().
+ */
+typedef struct circpad_distribution_t {
+  circpad_distribution_type_t type;
+  double param1;
+  double param2;
+} circpad_distribution_t;
+
+/**
+ * A circuit padding state machine state.
+ *
+ * This struct describes the histograms and parameters of a single
+ * state in the adaptive padding machine. Instances of this struct
+ * exist in global circpad machine definitions that come from torrc
+ * or the consensus, and are immutable.
+ */
+typedef struct circpad_state_t {
+  /** how long the histogram is (in bins). Histograms must have at least 2
+   *  bins (or 0, if iat distributions are in use). */
+  uint8_t histogram_len;
+  /** histogram itself: an array of uint16s of tokens, whose
+   * widths are exponentially spaced, in microseconds */
+  circpad_hist_bin_t histogram[CIRCPAD_MAX_HISTOGRAM_LEN];
+  /** total number of tokens */
+  uint32_t histogram_total;
+
+  /** Microseconds of the first bin of histogram, or base of iat dist */
+  circpad_delay_t start_usec;
+  /** The span of the histogram in seconds, used to calculate bin with.
+   *  For iat dist use, this is used as a max delay cap on the distribution. */
+  uint16_t range_sec;
+
+  /**
+   * The iat_dist is a parametrized way of encoding inter-packet delay
```

Comment:
This comment never explains what iat_dist actually means

Reply:
Addressed in e1805053a48db1b489268c6ac3be5e3792fed00d.